pkg/storegateway/indexheader/stream_binary_reader.go

```diff
@@ -388,3 +347,138 @@ func (r *StreamBinaryReader) Close() error {
 	r.factory.Stop()
 	return nil
 }
+
+func (r *StreamBinaryReader) sparseHeader() *indexheaderpb.Sparse {
+	sparseHeader := &indexheaderpb.Sparse{}
+	sparseHeader.Symbols = r.symbols.NewSparseSymbol()
+	sparseHeader.PostingsOffsetTable = r.postingsOffsetTable.NewSparsePostingOffsetTable()
+	return sparseHeader
+}
+
+type bucketAndDiskSparseHeaderLoader struct {
+	ctx       context.Context
+	bkt       objstore.Bucket
+	localPath string
+	logger    log.Logger
+}
+
+func (l bucketAndDiskSparseHeaderLoader) SparseHeader() (*indexheaderpb.Sparse, error) {
+	defer func(start time.Time) {
+		level.Debug(l.logger).Log("msg", "read and parsed sparse index header", "path", l.localPath, "elapsed", time.Since(start))
+	}(time.Now())
+
+	sparseData, existsLocally, err := l.sparseHeaderBytes()
+	if err != nil {
+		return nil, err
+	}
+	if len(sparseData) > 0 && !existsLocally {
+		err = l.persistHeaderToDisk(sparseData)
+		if err != nil {
+			level.Error(l.logger).Log("msg", "couldn't persist sparse index to disk; will download from object store next time too", "path", l.localPath, "err", err)
+			// otherwise ignore the error because loading the header is the priority; not persisting it to disk
+		}
+	}
+
+	sparseHeaders, err := l.parseSparseHeader(sparseData)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse sparse index-header: %w", err)
+	}
+	return sparseHeaders, nil
+}
+
+func (l bucketAndDiskSparseHeaderLoader) sparseHeaderBytes() (b []byte, existsLocally bool, _ error) {
```

Comment:
Might be useful to add a metric for the number of sparse index headers read from disk vs read from object storage vs recreated - if we saw a very high number of recreations, that might indicate some kind of issue